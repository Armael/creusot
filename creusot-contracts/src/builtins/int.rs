use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
