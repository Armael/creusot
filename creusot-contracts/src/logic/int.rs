use super::model::*;
use crate as creusot_contracts;
use crate::logic::*;
use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        absurd
    }
}

impl From<u8> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn from(_: u8) -> Self {
        absurd
    }
}
impl Model for u8 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        absurd
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: u64) -> Self {
        absurd
    }
}
impl Model for u64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn from(_: i64) -> Self {
        absurd
    }
}
impl Model for i64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        absurd
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        absurd
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        absurd
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
