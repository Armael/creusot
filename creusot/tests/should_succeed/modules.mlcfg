module Ambient
use Ref 
use mach.int.Int 
use mach.int.Int32
use mach.int.Int64
use mach.int.UInt32
use mach.int.UInt64
use string.Char
use floating_point.Single
use floating_point.Double
(** Generic Type for borrowed values *) 
type borrowed 'a = 
{ current : 'a ; 
final : 'a; (* The "future" value when borrow will end *) 
} 
let function ( *_ ) x = x.current 
let function ( ^_ ) x = x.final 
val borrow_mut (a : 'a) : borrowed 'a 
ensures { *result = a }
type usize = int
type isize = int
  scope Nested
    scope Further
      let cfg another () : bool
        =
      var _0 : bool;
      {
        goto BB0;
      }
      BB0 {
        _0 <- false;
        _0
      }

    end
    type nested  =
        | Test

    let cfg innerFunc () : bool
      ensures { result = true }
      =
    var _0 : bool;
    var _1 : nested;
    {
      goto BB0;
    }
    BB0 {
      _1 <- Test;
      _0 <- true;
      _0
    }

  end
  let cfg main () : ()
    =
  var _0 : ();
  var _1 : bool;
  var _2 : bool;
  {
    goto BB0;
  }
  BB0 {
    _1 <- Nested.innerFunc ();
    goto BB1
  }
  BB1 {
    _2 <- Nested.Further.another ();
    goto BB2
  }
  BB2 {
    _0 <- ();
    _0
  }

end
