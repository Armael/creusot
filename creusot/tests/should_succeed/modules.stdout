module Type
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type nested_nested  = 
    | Nested_Nested_Test
    
  predicate drop_nested_nested (self : nested_nested) = 
    let Nested_Nested_Test = self in true
end
module NestedInnerFunc
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  let rec cfg impl () : bool
    ensures { result = true }
    
   = 
  var _0 : bool;
  var _1 : Type.nested_nested;
  {
    goto BB0
  }
  BB0 {
    _1 <- Type.Nested_Nested_Test;
    assume { Type.drop_nested_nested _1 };
    _0 <- true;
    return _0
  }
  
end
module NestedFurtherAnother
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  let rec cfg impl () : bool = 
  var _0 : bool;
  {
    goto BB0
  }
  BB0 {
    _0 <- false;
    return _0
  }
  
end
module Main
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  clone NestedFurtherAnother as Another1
  clone NestedInnerFunc as InnerFunc0
  let rec cfg impl () : () = 
  var _0 : ();
  var _1 : bool;
  var _2 : bool;
  {
    goto BB0
  }
  BB0 {
    _1 <- InnerFunc0.impl ();
    goto BB1
  }
  BB1 {
    assume { drop_bool _1 };
    _2 <- Another1.impl ();
    goto BB2
  }
  BB2 {
    assume { drop_bool _2 };
    _0 <- ();
    return _0
  }
  
end
